### PR TITLE
improve DefaultClientDialContext

### PR DESCRIPTION
**Pull request by p0lyn0mial**
wrapping the DialContext method and setting the socket options on a copy of the file descriptor is not deterministic.
as per documentation attempting to change properties of the original using this duplicate
may or may not have the desired effect.

instead we should set the options on the original file descriptor.

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
I performed the manual tests described https://github.com/openshift/library-go/pull/926#issuecomment-717888348 and https://github.com/openshift/library-go/pull/926#issuecomment-718694746 and haven't found any issues. 

I got the same results.

File: pkg/network/dialer_linux.go [hunk 1]
```diff
@@ -16,28 +15,18 @@ func dialerWithDefaultOptions() DialContext {
 	nd := &net.Dialer{
 		// TCP_USER_TIMEOUT does affect the behaviour of connect() which is controlled by this field so we set it to the same value
 		Timeout: 25 * time.Second,
+		// KeepAlive must to be set to a negative value to stop std library from applying the default values
+		// by doing so we ensure that the options we are interested in won't be overwritten
+		KeepAlive: time.Duration(-1),
+		Control: func(network, address string, con syscall.RawConn) error {
+			var err error
+			err = con.Control(func(fd uintptr) {
+				err = setDefaultSocketOptions(int(fd))
```
Comment (sttts) [inline pkg/network/dialer_linux.go:24]: which err will survive?

Comment (p0lyn0mial) [reply pkg/network/dialer_linux.go:24]: /hold

it should have been `return setDefaultSocketOptions(int(fd))`

Comment (p0lyn0mial) [reply pkg/network/dialer_linux.go]: /hold cancel

**Pull request merged**